### PR TITLE
Remove abort steps when lock is granted

### LINKED ISSUE #79: Spec does not say what should happen when aborted after lock acquired
There is a WPT test for this: https://github.com/web-platform-tests/wpt/blob/81bfb439212b8001925d46617fe4a37d4bdc6498/web-locks/signal.tentative.https.any.js#L134-L166

But the spec for abort is:

https://wicg.github.io/web-locks/#dom-lockmanager-request

>If options’ signal dictionary member is present, then add the following abort steps to options’ signal dictionary member:
>1. Enqueue the steps to abort the request request to the lock task queue.
>2. Reject promise with an "AbortError" DOMException.

That seems like it always reject regardless of lock acquisition. Should the spec be fixed?

### PR DESCRIPTION
**Pull request by saschanaz**
Closes #79 

I tried describing https://github.com/WICG/web-locks/issues/79#issuecomment-942572758, which matches the currently proposed Gecko implementation.

***
<a href="https://pr-preview.s3.amazonaws.com/saschanaz/web-locks/pull/80.html" title="Last updated on Nov 1, 2021, 9:17 PM UTC (0797b65)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/web-locks/80/b3e696f...saschanaz:0797b65.html" title="Last updated on Nov 1, 2021, 9:17 PM UTC (0797b65)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -581,9 +578,10 @@ The <dfn attribute for=Lock>mode</dfn> getter's steps are to return the associat
 <!-- ====================================================================== -->
 
 <div algorithm>
-To <dfn>request a lock</dfn> with |promise|, |agent|, |clientId|, |origin|, |callback|, |name|, |mode|, |ifAvailable| and |steal|:
+To <dfn>request a lock</dfn> with |promise|, |agent|, |clientId|, |origin|, |callback|, |name|, |mode|, |ifAvailable|, |steal|, and |signal|:
 
-1. Let |request| be a new [=lock request=] (|agent|, |clientId|, |origin|, |name|, |mode|, |callback|, |promise|).
+1. Let |request| be a new [=lock request=] (|agent|, |clientId|, |origin|, |name|, |mode|, |callback|, |promise|, |signal|).
+1. If |signal| is present, then <a for=AbortSignal>add</a> the algorithm [=signal to abort the request=] |request| to |signal|.
```
Comment (inexorabletash) [inline index.bs:584]: You should be able to write this as `[=AbortSignal/add=]`.

File: index.bs [hunk 2]
```diff
@@ -656,10 +662,12 @@ To <dfn>process the lock request queue</dfn> |queue|:
         1. Let |mode| be |request|'s [=lock request/mode=].
         1. Let |callback| be |request|'s [=lock request/callback=].
         1. Let |p| be |request|'s [=lock request/promise=].
+        1. Let |signal| be |request|'s [=lock request/signal=].
         1. Let |waiting| be [=a new promise=].
         1. Let |lock| be a new [=lock-concept|lock=] with [=lock-concept/agent=] |agent|, [=lock-concept/clientId=] |clientId|, [=lock-concept/origin=] |origin|, [=lock-concept/mode=] |mode|, [=lock-concept/name=] |name|, [=lock-concept/released promise=] |p|, and [=lock-concept/waiting promise=] |waiting|.
         1. [=set/Append=] |lock| to |origin|'s [=held lock set=].
         1. [=Enqueue the following steps=] on |callback|'s [=relevant settings object=]'s [=responsible event loop=]:
+            1. If |signal| is present, then <a for=AbortSignal>remove</a> the algorithm [=signal to abort the request=] |request| from |signal|.
```
Comment (inexorabletash) [inline index.bs:670]: You should be able to write this as `[=AbortSignal/remove=]`.

Comment (inexorabletash) [inline index.bs:670]: Hmmm... Adding/removing should be on the event loop, as you've done here. But it seems like there's a potential race between aborting (e.g. calling abort(), which be within a task the event loop) and the lock being granted (on the lock task queue) that this doesn't solve. 

Before telling script that it was granted the lock, do we need to verify that it wasn't aborted and drop the lock if so?

Comment (saschanaz) [reply index.bs:670]: That sounds about right. I'll add that.

**Comment by inexorabletash** [comment]
@pwnall - can you take a look too?

File: index.bs [hunk 3]
```diff
@@ -656,10 +662,14 @@ To <dfn>process the lock request queue</dfn> |queue|:
         1. Let |mode| be |request|'s [=lock request/mode=].
         1. Let |callback| be |request|'s [=lock request/callback=].
         1. Let |p| be |request|'s [=lock request/promise=].
+        1. Let |signal| be |request|'s [=lock request/signal=].
         1. Let |waiting| be [=a new promise=].
         1. Let |lock| be a new [=lock-concept|lock=] with [=lock-concept/agent=] |agent|, [=lock-concept/clientId=] |clientId|, [=lock-concept/origin=] |origin|, [=lock-concept/mode=] |mode|, [=lock-concept/name=] |name|, [=lock-concept/released promise=] |p|, and [=lock-concept/waiting promise=] |waiting|.
         1. [=set/Append=] |lock| to |origin|'s [=held lock set=].
         1. [=Enqueue the following steps=] on |callback|'s [=relevant settings object=]'s [=responsible event loop=]:
+            1. If |signal| is present, then run these steps:
+                1. If |signal|'s [=AbortSignal/aborted flag=] is set, then return.
```
Comment (inexorabletash) [inline index.bs:671]: Does this leave the lock held, or will it be immediately dropped?

Comment (saschanaz) [reply index.bs]: Oops, sorry, I hope the new commit fix this.

File: index.bs [hunk 4]
```diff
@@ -668,7 +668,8 @@ To <dfn>process the lock request queue</dfn> |queue|:
         1. [=set/Append=] |lock| to |origin|'s [=held lock set=].
         1. [=Enqueue the following steps=] on |callback|'s [=relevant settings object=]'s [=responsible event loop=]:
             1. If |signal| is present, then run these steps:
-                1. If |signal|'s [=AbortSignal/aborted flag=] is set, then return.
```
Comment (inexorabletash) [inline index.bs]: This latest diff change dropped the **return** here - I think you still want to exit these overall steps early in this case, don't you?

Comment (saschanaz) [reply index.bs]: 😬😬😬😬 yes, will add it right now

**Comment by inexorabletash** [comment]
Thanks, and again sorry for the delay.

**Pull request merged**